### PR TITLE
feat(context): improve ctx.inspect() to allow classes with colliding names

### PR DESCRIPTION
**Pull request by raymondfeng**
Classes with the same name can be declared and bound from different scopes
or modules. For inspection, we need to have unique names to treat them
differently.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/context/src/context.ts [hunk 1]
```diff
@@ -797,27 +798,81 @@ export class Context extends EventEmitter {
...
+
+  visit(ctor: Constructor<unknown>) {
+    let name = this.classes.get(ctor);
+    if (name == null) {
+      name = ctor.name;
+      // Now check if the name collides with another class
+      let index = this.nameIndex.get(name);
+      if (typeof index === 'number') {
+        // A conflict is found, mangle the name
+        this.nameIndex.set(name, ++index);
```
Comment (emonddr) [inline packages/context/src/context.ts]: The resulting names look a lot like the names of nested java classes.  `Myprovider$1` .

For bindings, you gave them an unique id: `Binding_2_3`  for example.
Can we do the same?

Comment (raymondfeng) [reply packages/context/src/context.ts]: Binding ids are not generated by `inspect()`.

I was trying to use `_1` but there is a possibility that a class named `MyProvider_1` may exist. I don't want to make the name mangling too complex.

Comment (emonddr) [reply packages/context/src/context.ts]: I see ok.

File: packages/context/src/context.ts [hunk 2]
```diff
@@ -797,27 +798,81 @@ export class Context extends EventEmitter {
...
+        name = `${name}$${index}`;
```
Comment (bajtos) [inline packages/context/src/context.ts]: Can we use a space character as the separator? I believe spaces are not allowed in JavaScript class/function names.

```suggestion
        name = `${name} #${index}`;
```

(`Myprovider #1`, `Binding #2`, etc.)

**Pull request merged**